tests/flow/test_all_shortest_paths.py

```diff
@@ -0,0 +1,86 @@
+        redis_graph.add_edge(edge)
+
+        edge = Edge(nodes[4], "E2", nodes[3], properties={"weight": 1})
+        redis_graph.add_edge(edge)
+
+        edge = Edge(nodes[1], "E2", nodes[3], properties={"weight": 2})
+        redis_graph.add_edge(edge)
+
+        redis_graph.commit()
+
+    def test01_invalid_shortest_paths(self):
+        query = """MATCH (a {v: 1}), (b {v: 4}), p = shortestPath((a)-[*]->(b)) RETURN p"""
+        try:
+            redis_graph.query(query)
+            self.env.assertTrue(False)
+        except redis.exceptions.ResponseError as e:
+            self.env.assertIn("RedisGraph currently only supports shortestPath in WITH or RETURN clauses", str(e))
+
+    def test02_all_shortest_paths(self):
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((a)-[*]->(b)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        actual_result = redis_graph.query(query)
+        # The 2 2-hop traversals should be found
+        expected_result = [[[nodes[0], nodes[1], nodes[3]]],
+                           [[nodes[0], nodes[4], nodes[3]]]]
+        self.env.assertEqual(actual_result.result_set, expected_result)
+        # Verify that a right-to-left traversal produces the same results
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((b)<-[*]-(a)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        self.env.assertEqual(actual_result.result_set, expected_result)
+
+    def test03_all_shortest_min_hops(self):
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((a)-[*3..]->(b)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        actual_result = redis_graph.query(query)
+        # The 1 3-hop traversal should be found
+        expected_result = [[[nodes[0], nodes[1], nodes[2], nodes[3]]]]
+        self.env.assertEqual(actual_result.result_set, expected_result)
+        # Verify that a right-to-left traversal produces the same results
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((b)<-[*3..]-(a)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        self.env.assertEqual(actual_result.result_set, expected_result)
+
+    def test04_all_shortest_edge_filter(self):
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((a)-[* {weight: 1}]->(b)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        actual_result = redis_graph.query(query)
+        # The 1 2-hop traversal with all edge weights of 1 should be found
+        expected_result = [[[nodes[0], nodes[4], nodes[3]]]]
+        self.env.assertEqual(actual_result.result_set, expected_result)
+        # Verify that a right-to-left traversal produces the same results
+        query = """MATCH (a {v: 1}), (b {v: 4}) WITH a, b MATCH p = allShortestPaths((b)<-[* {weight: 1}]-(a)) RETURN nodes(p) AS nodes ORDER BY nodes"""
+        self.env.assertEqual(actual_result.result_set, expected_result)
```

Comment:
Generally speaking I think it would be easier to start with support for `allShortestPaths` only in WITH and RETURN clauses, is that a valid possibility ?

Reply:
The problem with this idea is that to support `allShortestPaths` in WITH and RETURN clauses, we would need to make it a new `ArithmeticExpression` function. Each of these functions is invoked once for each input and produces one output, so there's no way to directly return multiple paths connecting the same 2 nodes. We could return an array of all such paths, but then the format would no longer match Cypher.